File: backend/app/schemas/agentive/speech.py
"""Wire shapes for ``/agentive/speech/*`` — voice input for the agent chat."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.speech_preferences import SpeechPreferences, validate_language_tag


class SpeechEngineOption(BaseModel):
    """One recognizer the composer may use, in preference order."""

    engine: str
    source: Literal["workspace", "browser"]
    display_name: str
    streaming: bool = True
    privacy_note: Optional[str] = None


class SpeechProviderStatus(BaseModel):
    """The workspace's provider, as far as the caller needs to know."""

    configured: bool
    #: False for workspace guests even when a provider is configured.
    available_to_you: bool
    provider: Optional[str] = None
    model: Optional[str] = None
    source: Optional[Literal["byok", "platform"]] = None


class SpeechLimits(BaseModel):
    max_session_seconds: int
    session_ttl_seconds: int


class SpeechConfigResponse(BaseModel):
    workspace_id: str
    workspace_role: str
    preferences: SpeechPreferences
    engines: List[SpeechEngineOption]
    #: Engine the caller's preference points at; the browser still checks
    #: support (e.g. no Web Speech on Firefox) before offering it.
    preferred_engine: Optional[str]
    provider: SpeechProviderStatus
    limits: SpeechLimits


class SpeechSessionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    #: Overrides the caller's preferred language for this session.
    language: Optional[str] = Field(default=None, max_length=35)

    @field_validator("language")
    @classmethod
    def _language(cls, value: Optional[str]) -> Optional[str]:
        return None if value is None else validate_language_tag(value)


class SpeechSessionResponse(BaseModel):
    """A short-lived browser credential — never the provider API key."""

    engine: str
    transport: Literal["webrtc", "websocket"]
    connect_url: str
    client_secret: str
    expires_at: datetime
    params: Dict[str, Any] = Field(default_factory=dict)
    max_session_seconds: int

File: frontend/src/api/speech.ts
/**
 * Voice input API — effective config, session minting, per-user preferences.
 *
 * Hand-mirrors backend/app/schemas/agentive/speech.py and
 * backend/app/schemas/speech_preferences.py. Every call opts out of the
 * system notification bar: the composer and the settings panel report speech
 * errors inline, and a missing provider is a normal state, not an outage.
 */
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';

import { useScope } from '../context/ScopeContext';
import apiClient from './client';

export type SpeechEngineChoice = 'auto' | 'browser';
export type HotkeyMode = 'hold_or_toggle' | 'toggle' | 'hold';

export interface SpeechPreferences {
  enabled: boolean;
  engine: SpeechEngineChoice;
  /** `auto` or a BCP-47 tag such as `en-US`. */
  language: string;
  /** `Mod+Shift+Space`-style chord; `Mod` is ⌘ on macOS, Ctrl elsewhere. */
  hotkey: string;
  hotkey_mode: HotkeyMode;
  auto_send_on_stop: boolean;
  /** Seconds of silence before dictation stops; 0 disables. */
  silence_timeout_seconds: number;
}

export interface SpeechEngineOption {
  engine: string;
  source: 'workspace' | 'browser';
  display_name: string;
  streaming: boolean;
  privacy_note: string | null;
}

export interface SpeechProviderStatus {
  configured: boolean;
  available_to_you: boolean;
  provider: string | null;
  model: string | null;
  source: 'byok' | 'platform' | null;
}

export interface SpeechConfig {
  workspace_id: string;
  workspace_role: string;
  preferences: SpeechPreferences;
  /** Recognizers the caller may use, workspace provider first. */
  engines: SpeechEngineOption[];
  preferred_engine: string | null;
  provider: SpeechProviderStatus;
  limits: { max_session_seconds: number; session_ttl_seconds: number };
}

/** Short-lived browser credential — never the provider API key. */
export interface SpeechSessionResponse {
  engine: string;
  transport: 'webrtc' | 'websocket';
  connect_url: string;
  client_secret: string;
  expires_at: string;
  params: Record<string, unknown>;
  max_session_seconds: number;
}

const QUIET = { __suppressSystemNotify: true } as never;

export const SPEECH_CONFIG_QUERY_ROOT = ['speech', 'config'] as const;
export const SPEECH_PREFS_KEY = ['speech', 'preferences'] as const;

export async function fetchSpeechConfig(): Promise<SpeechConfig> {
  const { data } = await apiClient.get<SpeechConfig>('/agentive/speech/config', QUIET);
  return data;
}

export async function mintSpeechSession(
  body: { language?: string } = {},
): Promise<SpeechSessionResponse> {
  const { data } = await apiClient.post<SpeechSessionResponse>(
    '/agentive/speech/session',
    body,
    QUIET,
  );
  return data;
}

export async function fetchSpeechPreferences(): Promise<SpeechPreferences> {
  const { data } = await apiClient.get<SpeechPreferences>(
    '/users/me/speech-preferences',
    QUIET,
  );
  return data;
}

export async function patchSpeechPreferences(
  partial: Partial<SpeechPreferences>,
): Promise<SpeechPreferences> {
  const { data } = await apiClient.patch<SpeechPreferences>(
    '/users/me/speech-preferences',
    partial,
    QUIET,
  );
  return data;
}

/** Effective config for the active workspace; refetches on workspace switch. */
export function useSpeechConfig() {
  const { activeWorkspace } = useScope();
  return useQuery<SpeechConfig>({
    queryKey: [...SPEECH_CONFIG_QUERY_ROOT, activeWorkspace?.id ?? null],
    queryFn: fetchSpeechConfig,
    staleTime: 60_000,
    retry: false,
  });
}

export function useSpeechPreferences() {
  return useQuery<SpeechPreferences>({
    queryKey: SPEECH_PREFS_KEY,
    queryFn: fetchSpeechPreferences,
  });
}

export function useUpdateSpeechPreferences() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: patchSpeechPreferences,
    onSuccess: data => {
      qc.setQueryData(SPEECH_PREFS_KEY, data);
      void qc.invalidateQueries({ queryKey: SPEECH_CONFIG_QUERY_ROOT });
    },
  });
}
